docs/language/learn-ql/javascript/type-tracking.rst

```diff
@@ -0,0 +1,523 @@
+so we'll use type tracking again in order to take the receiver type into account.
+
+The receiver ``snapshot`` is a parameter to a callback function, which ultimately escapes
+into the ``once()`` call. We'll extend our model from above to use back-tracking to find
+all functions that flow into the ``once()`` call. Type tracking backwards is not much
+different from forwards; the differences are:
+
+- The ``TypeTracker`` parameter instead has type ``TypeBackTracker``.
+- The call to ``.track()`` is instead a call to ``.backtrack()``
+- To ensure the initial value is a source node, a call to ``getALocalSource()`` is usually required.
+
+.. code-block:: ql
+
+  SourceNode firebaseSnapshotCallback(string refName, TypeBackTracker t) {
+    t.start() and
+    result = firebaseRef(refName).getAMethodCall("once").getArgument(1).getALocalSource()
+    or
+    exists(TypeBackTracker t2 |
+      result = firebaseSnapshotCallback(refName, t2).backtrack(t2, t)
+    )
+  }
+
+  FunctionNode firebaseSnapshotCallback(string refName) {
+    result = firebaseSnapshotCallback(refName, TypeBackTracker::end())
+  }
+
+Now, ``firebaseSnapshotCallback("forecast")`` finds the function being passed to ``fetchForecast``.
+Based on that we can track the ``snapshot`` value and find the ``val()`` call itself:
+
+.. code-block:: ql
+
+  SourceNode firebaseSnapshot(string refName, TypeTracker t) {
+    t.start() and
+    result = firebaseSnapshotCallback(refName).getParameter(0)
+    or
+    exists(TypeTracker t2 |
+      result = firebaseSnapshot(refName, t2).track(t2, t)
+    )
+  }
+
+  SourceNode firebaseSnapshot(string refName) {
+    result = firebaseSnapshot(refName, TypeTracker::end())
+  }
+
+  MethodCallNode firebaseDatabaseRead(string refName) {
+    result = firebaseSnapshot(refName).getAMethodCall("val")
+  }
+
+With this addition, ``firebaseDatabaseRead("forecast")`` finds the call to ``snapshot.val()`` which contains the value of the forecast.
+
+`Here <https://lgtm.com/query/8761360814276109092>`__ is a run of an example query using the model to find `val` calls.
+
+Summary
+-------
+
+This covers the use of the type tracking library. To recap, use this template to define forward type tracking predicates:
+
+.. code-block:: ql
+
+  SourceNode myType(TypeTracker t) {
+    t.start() and
+    result = /* SourceNode to track */
+    or
+    exists(TypeTracker t2 |
+      result = myType(t2).track(t2, t)
+    )
+  }
+
+  SourceNode myType() {
+    result = myType(TypeTracker::end())
+  }
+
+Use this template to define backward type tracking predicates:
+
+.. code-block:: ql
+
+  SourceNode myType(TypeBackTracker t) {
+    t.start() and
+    result = (/* argument to track */).getALocalSource()
+    or
+    exists(TypeBackTracker t2 |
+      result = myType(t2).backtrack(t2, t)
+    )
+  }
+
+  SourceNode myType() {
+    result = myType(TypeBackTracker::end())
+  }
+
+Note that these predicates all return ``SourceNode``,
```

Comment:
Should we mention that the type tracking library also supports tracking non-source nodes?

Reply:
Added a link to `smallstep` and added examples to the QLDoc for the those predicates.